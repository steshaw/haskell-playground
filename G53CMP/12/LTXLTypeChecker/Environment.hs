-- ***************************************************************************
-- *									     *
-- *		     Less Trivial eXpression Language (LTXL)		     *
-- *									     *
-- *	Module:		Environment					     *
-- *	Purpose:	Type-checking environment and operations	     *
-- *	Author:		Henrik Nilsson					     *
-- *									     *
-- *           Example for G53CMP, lectures 12 & 13, November 2011           *
-- *									     *
-- ***************************************************************************

module Environment (
    VarAttr,
    Env,	-- Abstract
    glblLvl,    -- :: Int
    initEnv,    -- :: [(Id, Type)] -> [(UnOp, Type)] -> [(BinOp, Type)] -> Env
    enterVar,	-- :: Id -> Int -> Type -> Env -> Either Env ErrorMsg
    lookupVar,	-- :: Id -> Env -> Either VarAttr ErrorMsg
    lookupUO,	-- :: UnOp -> Env -> Type
    lookupBO	-- :: BinOp -> Env -> Type
) where

import List (nub)

import Diagnostics
import AbstractSyntax (Id, UnOp, BinOp)
import Type (Type)


------------------------------------------------------------------------------
-- Environment (Symbol table)
------------------------------------------------------------------------------

-- The environment stores information about entities, in particular their
-- type.
--
-- There are three kinds of entities. Each have a distinct kind of name
-- (used for looking it up), and a distinct set of attributes:
--   * Variables
--        Name: Identifier
--        Attributes: Scope level and type.
--   * Unary operator
--        Name: UnOp
--        Attributes: Type only (the scope level is always the global one).
--   * Binary operator
--        Name: BinOp
--        Attributes: Type only (the scope level is always the global one).
--
--
-- The reason that operators are separated from variables is that the LTXL
-- syntax defines a fixed set of operators (user-defined operators are not
-- supported). For the same reason, we also know that an operator is an entity
-- defined at the global level.
-- 
-- The environment is an Abstract Data Type (ADT) which is achieved by not
-- exporting its constructor from this module. It is split into three parts,
-- one for each kind of entity. For the sake of simplicity, we use a linear
-- association list to represent the environment. By prepending new
-- declarations to the list, and searching from the beginning, it is ensured
-- that we will always find an identifier in the closest containing scope, as
-- specified by the LTXL Scope Rules.


type VarAttr = (Int, Type)

data Env = Env {
               varEnv :: [(Id, VarAttr)],	-- The variable environment
               uoEnv  :: [(UnOp, Type)],	-- The unary operator env.
               boEnv  :: [(BinOp, Type)]	-- Tye binary operator env.
           } 


glblLvl :: Int
glblLvl = 0


-- Creates an initial environment (everything defined at the top level).

initEnv :: [(Id, Type)] -> [(UnOp, Type)] -> [(BinOp, Type)] -> Env
initEnv ve uoe boe =
    if ok then
        Env {
            varEnv = [ (i, (glblLvl, tp)) | (i, tp) <- ve ],
            uoEnv  = uoe,
            boEnv  = boe
        }
    else
        internalError "Bad initial environment: multiply defined entities"
    where
        ok = length is == length (nub is)
             && length uos == length (nub uos)
             && length bos == length (nub bos)
        is  = map fst ve
        uos = map fst uoe
        bos = map fst boe


-- Enters a variable at the given scope level and of the given type into the
-- environment. A check is first performed so that the no other variable with
-- the same name has been defined at the same scope level. If not, the new
-- variable is entered into the table. Otherwise an error message is returned.

enterVar :: Id -> Int -> Type -> Env -> Either Env ErrorMsg
enterVar i l t env@(Env {varEnv = ve})
    | not (isDefined i l ve) = Left (env {varEnv = ((i,(l,t)) : ve)}) 
    | otherwise              = Right errMsg
    where
        isDefined i l [] = False
        isDefined i l ((i',(l',_)) : ve)
            | l < l'  = error "Should not happen! (Bad level.)"
            | l > l'  = False
            | i == i' = True
            | otherwise = isDefined i l ve

        errMsg = "Variable " ++ i
                 ++ " is already defined at this level ("
                 ++ show l ++ ")."


-- Looks up a variable and returns its attributes if found.
-- Otherwise returns an error message.

lookupVar :: Id -> Env -> Either VarAttr ErrorMsg
lookupVar i env = lvRec i (varEnv env)
    where
        lvRec i [] = Right ("Variable " ++ i ++ " is not defined.")
        lvRec i ((i',a) : ve)
            | i == i'   = Left a
            | otherwise = lvRec i ve


-- Looks up a unary operator and returns its type.
-- It should always be found, otherwise there is an error in the compiler.

lookupUO :: UnOp -> Env -> Type
lookupUO uo env =
    case lookup uo (uoEnv env) of
        Just tp -> tp
        Nothing -> internalError ("Unknown unary operator " ++ show uo)


-- Looks up a binary operator and returns its type.
-- It should always be found, otherwise there is an error in the compiler.

lookupBO :: BinOp -> Env -> Type
lookupBO bo env =
    case lookup bo (boEnv env) of
        Just tp -> tp
        Nothing -> internalError ("Unknown unary operator " ++ show bo)
