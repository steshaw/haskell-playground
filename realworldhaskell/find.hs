--
-- Adapted from RWH Chapter 9 "I/O Case Study: A Library for Searching the Filesystem"
--

module Main where

import System.Environment (getArgs)
import Control.Monad (forM)
import System.Directory (doesDirectoryExist, getDirectoryContents)
import System.FilePath ((</>))

find :: FilePath -> IO [FilePath]
find topDir = do
  names <- getDirectoryContents topDir
  let properNames = filter (`notElem` [".", ".."]) names
  paths <- forM properNames $ \ name -> do
    let path = topDir </> name
    isDirectory <- doesDirectoryExist path
    if isDirectory then find path else return [path]
  return (concat paths)

main = do
  getArgs >>= \args -> 
    case args of
      [path] -> find path >>= mapM_ putStrLn
