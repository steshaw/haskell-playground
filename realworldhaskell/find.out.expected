..
../wc.hs
../sieve_of_eratothenes.hs
../arrows.hs
../repeatN.hs
../factorial.hs
../monads.hs
../listToString.hs
../.gitignore
../realworldhaskell
../realworldhaskell/toUpperBanner.o
../realworldhaskell/Steshaw.hi
../realworldhaskell/load-all
../realworldhaskell/fixeols.o
../realworldhaskell/tempFileFun.hi
../realworldhaskell/SplitLines.o
../realworldhaskell/readDouble.hs
../realworldhaskell/Tree.hs
../realworldhaskell/fixeols.hs
../realworldhaskell/toupper.o
../realworldhaskell/tempFileFun.hs
../realworldhaskell/find.o
../realworldhaskell/toUpperBanner.hi
../realworldhaskell/SplitLines.hs
../realworldhaskell/ZippityChallenge.hs
../realworldhaskell/Steshaw.hs
../realworldhaskell/find.out.actual
../realworldhaskell/readDouble.o
../realworldhaskell/InteractWith.hi
../realworldhaskell/copy.hs
../realworldhaskell/copy.hi
../realworldhaskell/Makefile
../realworldhaskell/InteractWith.hs
../realworldhaskell/InteractWith.o
../realworldhaskell/find.hs
../realworldhaskell/genProgs
../realworldhaskell/copy.o
../realworldhaskell/input.txt
../realworldhaskell/Chapter04.hs
../realworldhaskell/find.out.expected
../realworldhaskell/find
../realworldhaskell/toUpperBanner.hs
../realworldhaskell/Steshaw.o
../realworldhaskell/Chapter3.hs
../realworldhaskell/SplitLines.hi
../realworldhaskell/tempFileFun.o
../realworldhaskell/toupper.hs
../realworldhaskell/toupper.hi
../realworldhaskell/fixeols.hi
../realworldhaskell/readDouble.hi
../realworldhaskell/find.hi
../learnyouahaskell
../learnyouahaskell/copyFile.o
../learnyouahaskell/girlfriend.up.txt
../learnyouahaskell/lib.hs
../learnyouahaskell/newtypeVerseDataLaziness.o
../learnyouahaskell/p.o
../learnyouahaskell/colours.hi
../learnyouahaskell/bytestrings.hs
../learnyouahaskell/upcase.hs
../learnyouahaskell/class.hs
../learnyouahaskell/p.hi
../learnyouahaskell/foreverYoung.hi
../learnyouahaskell/newtype.hs
../learnyouahaskell/helloYou.hi
../learnyouahaskell/cat.o
../learnyouahaskell/copyFile.hi
../learnyouahaskell/catGirlfriend.hs
../learnyouahaskell/girlfriend.txt
../learnyouahaskell/newtypeVerseDataLaziness.hi
../learnyouahaskell/todo.hi
../learnyouahaskell/helloFirstLast.o
../learnyouahaskell/upGirlfriend.hs
../learnyouahaskell/todo.txt
../learnyouahaskell/foldable.hs
../learnyouahaskell/foreverYoung.o
../learnyouahaskell/reverseLoop.hi
../learnyouahaskell/palindromes.hi
../learnyouahaskell/upGirlfriend.o
../learnyouahaskell/tree.hs
../learnyouahaskell/Makefile
../learnyouahaskell/shortLines.o
../learnyouahaskell/todo.o
../learnyouahaskell/newtypeVerseDataLaziness.hs
../learnyouahaskell/colours.hs
../learnyouahaskell/shortLines.txt
../learnyouahaskell/p.hs
../learnyouahaskell/helloYou.hs
../learnyouahaskell/hello.hi
../learnyouahaskell/upcase.o
../learnyouahaskell/echo.hs
../learnyouahaskell/echo.o
../learnyouahaskell/copyFile.hs
../learnyouahaskell/beustSequence.hs
../learnyouahaskell/random.hs
../learnyouahaskell/helloFirstLast.hs
../learnyouahaskell/colours.o
../learnyouahaskell/reverseLoop.o
../learnyouahaskell/upGirlfriend.hi
../learnyouahaskell/reverseLoop.hs
../learnyouahaskell/trueish.hs
../learnyouahaskell/hello.o
../learnyouahaskell/catGirlfriend.o
../learnyouahaskell/upcase.hi
../learnyouahaskell/helloFirstLast.hi
../learnyouahaskell/catGirlfriend.hi
../learnyouahaskell/cat.hi
../learnyouahaskell/shortLines.hi
../learnyouahaskell/cons.hs
../learnyouahaskell/palindromes.hs
../learnyouahaskell/hello.hs
../learnyouahaskell/foreverYoung.hs
../learnyouahaskell/palindromes.o
../learnyouahaskell/helloYou.o
../learnyouahaskell/echo.hi
../learnyouahaskell/shortLines.hs
../learnyouahaskell/todo.hs
../learnyouahaskell/baby.hs
